src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,158 @@
+                indexedNode = indexedNode.next;
+            }
+        } else {
+            indexedNode = tail;
+            for (int i = size - 1; i > index; i--) {
+                indexedNode = indexedNode.prev;
+            }
+        }
+        return indexedNode;
+    }
+
     @Override
     public void add(T value) {
+        Node<T> checkTail = tail;
+        Node<T> newNode = new Node<>(checkTail, value, null);
+        tail = newNode;
+        if (checkTail == null) {
+            head = newNode;
+        } else {
+            checkTail.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index >= 0 && index <= size) {
+            if (index == size) {
+                add(value);
+            } else {
+                Node<T> next = getIndexedNode(index);
+                Node<T> prev = next.prev;
+                Node<T> newNode = new Node<>(prev, value, next);
+                next.prev = newNode;
+                if (prev == null) {
+                    head = newNode;
+                } else {
+                    prev.next = newNode;
+                }
+                size++;
+            }
+        } else {
+            throw new IndexOutOfBoundsException("Can't add value. Pass correct input index");
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= 0 && index < size) {
+            return getIndexedNode(index).item;
+        }
+        throw new IndexOutOfBoundsException("Can't get item. Pass correct input index");
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index >= 0 && index < size) {
+            Node<T> targetNode = getIndexedNode(index);
+            T oldValue = targetNode.item;
+            targetNode.item = value;
+            return oldValue;
+        }
+        throw new IndexOutOfBoundsException("Can't set value. Pass correct input index");
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= 0 && index < size) {
+            return unlink(getIndexedNode(index));
+        }
+        throw new IndexOutOfBoundsException("Can't remove item. Pass correct input index");
     }
 
     @Override
     public boolean remove(T object) {
+        if (object == null) {
+            for (Node<T> node = head; node != null; node = node.next) {
+                if (node.item == null) {
+                    unlink(node);
+                    return true;
+                }
+            }
+        } else {
+            for (Node<T> node = head; node != null; node = node.next) {
+                if (object.equals(node.item)) {
+                    unlink(node);
+                    return true;
+                }
+            }
+        }
```

Comment:
you can refactor and get rid of this ugly if else construction only by writing proper checks
```suggestion
        for (Node<T> node = head; node != null; node = node.next) {
            if ( /* new check here */ ) {
                unlink(node);
                return true;
            }
        }
        return false;
```

Reply:
if you feel stuck: check equals hashcode lessons once more, or just text me